hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala

```diff
@@ -0,0 +1,263 @@
+    s" destination=$destPath;" +
+    s" dynamicPartitionOverwrite=$dynamicPartitionOverwrite")
+
+  import PathOutputCommitProtocol._
+
+  /**
+   * Set up the committer.
+   * This creates it by talking directly to the Hadoop factories, instead
+   * of the V1 `mapred.FileOutputFormat` methods.
+   * @param context task attempt
+   * @return the committer to use. This will always be a subclass of
+   *         [[PathOutputCommitter]].
+   */
+  override protected def setupCommitter(context: TaskAttemptContext): PathOutputCommitter = {
+
+    logDebug(s"Setting up committer for path $destination")
+    committer = PathOutputCommitterFactory.createCommitter(destPath, context)
+
+    // Special feature to force out the FileOutputCommitter, so as to guarantee
+    // that the binding is working properly.
+    val rejectFileOutput = context.getConfiguration
+      .getBoolean(REJECT_FILE_OUTPUT, REJECT_FILE_OUTPUT_DEFVAL)
+    if (rejectFileOutput && committer.isInstanceOf[FileOutputCommitter]) {
+      // the output format returned a file output format committer, which
+      // is exactly what we do not want. So switch back to the factory.
+      val factory = PathOutputCommitterFactory.getCommitterFactory(
+        destPath,
+        context.getConfiguration)
+      logDebug(s"Using committer factory $factory")
+      committer = factory.createOutputCommitter(destPath, context)
+    }
+
+    logDebug(s"Using committer ${committer.getClass}")
+    logDebug(s"Committer details: $committer")
+    if (committer.isInstanceOf[FileOutputCommitter]) {
+      require(!rejectFileOutput,
+        s"Committer created is the FileOutputCommitter $committer")
+
+      if (committer.isCommitJobRepeatable(context)) {
+        // If FileOutputCommitter says its job commit is repeatable, it means
+        // it is using the v2 algorithm, which is not safe for task commit
+        // failures. Warn
+        logDebug(s"Committer $committer may not be tolerant of task commit failures")
+      }
+    }
+    committer
+  }
+
+  /**
+   * Create a temporary file for a task.
+   *
+   * @param taskContext task context
+   * @param dir         optional subdirectory
+   * @param ext         file extension
+   * @return a path as a string
+   */
+  override def newTaskTempFile(
+      taskContext: TaskAttemptContext,
+      dir: Option[String],
+      ext: String): String = {
+
+    val workDir = committer.getWorkPath
+    val parent = dir.map {
+      d => new Path(workDir, d)
+    }.getOrElse(workDir)
+    val file = new Path(parent, buildFilename(taskContext, ext))
+    logDebug(s"Creating task file $file for dir $dir and ext $ext")
+    file.toString
+  }
+
+  /**
+   * Request a path for new output file to be committed in job
+   * commit into the absolute path specified.
+   *
+   * The created file will be committed through the `rename()` operation
+   * of the filesystem/filestore connector, which may be a slow `O(data)`
```

Comment:
if I understand this right, then even after this change, all writes into a partitioned table will still require a slow rename?  Seems that way from my quick reading of this part of spark's code:

https://github.com/apache/spark/blob/b8e13b0aeae4a25e54a4c4b8a97bc2c19edd37bb/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L228-L235

is handling that not possible with spark's internal protocol?  Its fine we don't handle that for now, but should call it out in commit msg & file a follow up jira

Reply:
this is only for those cases where the protocol wants an absolute path. Traditionally spark cast the committer to a FileOutputCommitter and used methods there. Rather than make changes to what is a scarily complex piece of critical code to support new plugin committers, I just pulled up the relevant method into a new interface which can be used instead. 
Normally work goes into the subdirectory of the output, and even when partitioned there all is well. The PartitionedCommitter Ryan Blue wrote even does conflict handling within the partitions, so you can run a job targeting an existing output directory, and have it address conflict not at the base "fail if dest exists", but act on the final leaves of the partition with "rm * in that final dir", "fail if that final dir isn't empty" or "just add things relying on UUID insertion to avoid conflict and S3 update inconsistencies"

Happy to explore this more in future. FWIW, I think spark should be able to query a committer to see if could recover from a failure/partition during commit and then fail a job if a committer "went away" during that process. The existing MRv2 committer would return false here, as should the new EMR one. MRv1 and these S3A committers are resilient to failure here ... which is exactly what spark and Hadoop MR actually assume.

Reply:
ok I took another look at this, and the key part that I was missing in the code I mentioned above is that it only matters when `customPartitionLocations` is defined.  I'd still like to understand that case better, but agree its not a show stopper.